# Running CARLA simulator
This tutorial shows how to run and visualize CARLA on colab.



<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/MichaelBosello/carla-colab/blob/master/carla-simulator.ipynb">
    <img src="https://www.tensorflow.org/images/colab_logo_32px.png" />
    Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/MichaelBosello/carla-colab">
    <img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />
    View source on GitHub</a>
  </td>
</table>

<img src="https://raw.githubusercontent.com/MichaelBosello/carla-colab/master/img/screen.png" width="750">

---
---
---

## Introduction
[Carla](https://github.com/carla-simulator/carla) is an open-source simulator for autonomous driving research.

Since it is hardware demanding, it could be useful for students to run it on colab.

---

## Acknowledgements

Thanks to *Tomohiro*  for developing the code that allows remote access to colab.
Check out his [page](https://internet-of-tomohiro.netlify.app/google_colab/vnc.en.html) and [GitHub](https://github.com/demotomohiro/remocolab).

---

## Update

Ngrok doesn't work anymore with Google Colab (see [this issue](https://github.com/demotomohiro/remocolab/issues/79)). The tutorial has been updated to use Argo tunnel instead.

---


## Requirements



*   ~~[Ngrok](https://ngrok.com/) account (you don't need to buy premium plans).~~
*   [Install cloudflared](https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/install-and-setup/installation) on your PC.
*   TurboVNC Viewer. [Download here](https://sourceforge.net/projects/turbovnc/). (RealVNC and Microsoft Remote Desktop __don't__ work).
*   Ssh client (already present in all recent os).


## Change the runtime
In colab: click "Runtime" in the top bar -> "Change runtime type" -> change hardware accelerator to GPU.

## Installation & setup of remote access



1.   Run the snippet below
2.   ~~It will ask for your ngrok authtoken.~~

  *   ~~login to ngrok -> from the left menu: Authentication>Your Authroken -> copy~~
  *   ~~Enter it~~

3. ~~Select ngrok region (the region closest to you should be the fastest)~~
4. Wait the end of the process
  * It sets up OpenSSH server, VirtualGL and TurboVNC.
  * Then, access passwords are shown



In [1]:
!pip install git+https://github.com/demotomohiro/remocolab.git
import remocolab
remocolab.setupVNC()

  Cloning https://github.com/demotomohiro/remocolab.git to /tmp/pip-req-build-7yhyam3c
  Running command git clone --filter=blob:none --quiet https://github.com/demotomohiro/remocolab.git /tmp/pip-req-build-7yhyam3c
  Resolved https://github.com/demotomohiro/remocolab.git to commit 176445b2ca07cacf3a41036cf44bb38caf2c5049
  Preparing metadata (setup.py) ... done
  Created wheel for remocolab.py: filename=remocolab.py-0.1-py3-none-any.whl size=7526 sha256=6774ee74d37c784ba040df8f897e62825deb1f05f1c90167b25bfe9f5cf66bf9
  Stored in directory: /tmp/pip-ephem-wheel-cache-l29qtwqu/wheels/09/ff/ea/1c81a850640bbcd64d5f94d3d4143133d97363fa916f024fda
Successfully built remocolab.py
---
As ngrok doesn't work on colab for long time, default tunnel method has been changed to Argo tunnel.
Please read this for more details:
https://github.com/demotomohiro/remocolab/blob/master/README.md


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install openssh-server


Label(value='')

FloatProgress(value=0.0, layout=Layout(border='1px solid #118800'), max=1.0)

Install xfce4
Install xfce4-terminal
ED25519 key fingerprint of host:
256 SHA256:r1tMB43syydnDBNmSPLZFCUUSmxb1XBIhCxuiIg4U04 root@b6e609c462df (ED25519)
+--[ED25519 256]--+
|      ..oo*O*+.  |
|  E    =+B++...  |
|.= . . +=+O .    |
|= o . . ++ o     |
| o     .S = .    |
|         = *     |
|          B =    |
|         o =     |
|        o.       |
+----[SHA256]-----+

✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
root password: EVR4Fnc9oOv1M8EMYn5qZ-9cmktCtc92oZW15mcRZU0
colab password: _Sdo4qCFw_M-K8mdgcuOqgAtpDUKxFewnlSD5XKHFOQ
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
---
Execute following command on your local machine and login before running TurboVNC viewer:
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -oProxyCommand="cloudflared access ssh --hostname %h" -L 5901:localhost:5901 colab@floyd-remarks-proof-managers.trycloudflare.com
✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️✂️
VNC password: cJVydrnj
VNC vie

## Connect TurboVNC

1. Open a ssh connection to colab (It is necessary to secure VNC connection)
  * Execute the ssh command showed above
  * Use the "colab password" displayed above
2. Run TurboVNC Viewer
  * The connection address is `localhost:1`
  * Use the "VNC password" displayed above


 if everything worked properly, you will see a screen.

## Download CARLA

If you need another [version](https://github.com/carla-simulator/carla/blob/master/Docs/download.md) of CARLA, change the link below. You must choose a Linux version.



In [1]:
! sudo -u colab wget -O /home/colab/CARLA.tar.gz https://carla-releases.s3.eu-west-3.amazonaws.com/Linux/CARLA_0.9.10.tar.gz

sudo: unknown user colab
sudo: error initializing audit plugin sudoers_audit


## Extract the zip

In [ ]:
! sudo -u colab mkdir /home/colab/carla
! sudo -u colab tar -xf /home/colab/CARLA.tar.gz -C /home/colab/carla

## Setup PythonAPI

The PythonAPI in the zip doesn't work with the current Colab Python version. If you change the Carla version, if you install another Python version, or if Google updates Colab, then you may not need to execute the following commands (or you may need to find another egg file accordingly).

In [ ]:
! sudo -u colab rm -rf /home/colab/carla/PythonAPI/carla/dist/*.egg

In [ ]:
! sudo -u colab wget https://github.com/MichaelBosello/carla-colab/raw/master/carla-0.9.10-py3.6-linux-x86_64.egg -P /home/colab/carla/PythonAPI/carla/dist/

## Run CARLA

In [ ]:
! sudo -u colab DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh

# Run client

**You can run the client on colab or on your remote pc**

---

## Client on colab
Run the following commands in a terminal on the remote desktop

### Install requirement

1. `sudo pip3 install pygame`

  * The password for colab user is shown above

### Run the 'manual control' example

2. `cd /home/colab/carla/PythonAPI/examples`

3. `python3 manual_control.py`

---

## Client on your pc

Perform the following steps in your pc

### Install requirements

1. [install Carla](https://carla.readthedocs.io/en/latest/start_quickstart/#a-debian-carla-installation) (use the same version of the server)

2. `pip3 install pygame`

### Port forwarding

You have to forward two ports in addition to the one used for VNC

Copy the ssh command used for the VNC connection and run it with the port changed (change 5901 which is the VNC port). You have to run it two times (in two additional terminals) with those two ports: 2000 and 2001.

**For example (note that you have to change YOUR_ISTANCE_PORT to the port specified in your command):**

`ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -p YOUR_ISTANCE_PORT -L 2000:localhost:2000 colab@2.tcp.eu.ngrok.io
`

`ssh -o UserKnownHostsFile=/dev/null -o VisualHostKey=yes -p YOUR_ISTANCE_PORT -L 2001:localhost:2001 colab@2.tcp.eu.ngrok.io
`

### Run the 'manual control' example

1. `cd /opt/carla-simulator/PythonAPI/examples`

2. `python3 manual_control.py`

---
---
---

# Extra




## Display smooth animation
Open TurboVNC Viewer options and change "Encoding method" to "Tight + Medium-Quality JPEG" or "Tight + Low-Quality JPEG(WAN)". Some noises could appear but it will look smoother.

---
## Tips

### Keeping the virtual machine alive
You may want to use [this Chrome extension](https://chrome.google.com/webstore/detail/colab-alive/eookkckfbbgnhdgcbfbicoahejkdoele).

### How to transfer code
Options:
* Develop on your pc and use git to trasfer the code
* Develop on your pc and mount your Google Drive from colab (see snippet below)
* Install a code editor in the remote desktop (remember to save your code often on drive)

### Training

Save the checkpoints of your model on your Google Drive (see snippet below)



In [ ]:
from google.colab import drive
drive.mount('/content/drive')



---

## Notes



### CARLA command

The command to run the carla server is:

`DISPLAY=:1 vglrun /home/colab/carla/CarlaUE4.sh`

If you want to run the CARLA server in Python, you must use `subprocess.Popen` with the above command

### Users and terminals

The colab web-page runs commands as root. To execute them as a normal user I used `sudo -u colab`. This is necessary beacause __Unreal Engine can't be run as root__.

If you prefer, you can use the same commands in the remote Desktop or ssh terminal without `sudo -u colab`.

Python clients works only from the remote desktop or ssh client, and not from colab web-page.

### Check OpenGL

if you want to check the openGL implementation you can run in the remote desktop `vglrun /opt/VirtualGL/bin/glxinfo | grep NVIDIA`

You should see something like:

```
OpenGL vendor string: NVIDIA Corporation
OpenGL core profile version string: 4.6.0 NVIDIA 418.67
OpenGL core profile shading language version string: 4.60 NVIDIA
OpenGL version string: 4.6.0 NVIDIA 418.67
OpenGL shading language version string: 4.60 NVIDIA
```

If you get errors, it could be the VNC viewer (again, use TurboVNC).

### Closing Carla
Close CARLA always from the remote desktop, if you interrupt the execution from the colab web-page you will interrupt also the VNC server.